src/crypto/DeviceList.js

```diff
@@ -780,13 +783,17 @@ class DeviceListUpdateSerialiser {
             opts.token = this._syncToken;
         }
 
-        this._baseApis.downloadKeysForUsers(
-            downloadUsers, opts,
-        ).then(async (res) => {
-            const dk = res.device_keys || {};
-            const masterKeys = res.master_keys || {};
-            const ssks = res.self_signing_keys || {};
-            const usks = res.user_signing_keys || {};
+        const promises = [];
+        for (let i = 0; i < downloadUsers.length; i += this._deviceList._keyDownloadChunkSize) {
+            const userSlice = downloadUsers.slice(i, i + this._deviceList._keyDownloadChunkSize);
+            promises.push(this._baseApis.downloadKeysForUsers(userSlice, opts));
```

Comment:
our next problem is probably going to be rate limiting. In large rooms (>1000 users), this will be only 4 requests, but in massive rooms (>5000 users) this would be 20. Servers tend to allow a burst of up to 5 requests, with later ones being rate limited. We should parallelize at most 2-3 requests (probably 3?) to avoid a rate limit problem.

Reply:
Ah, good call. Do you want this to be implemented just here for the key downloads? Or would it make sense to push the rate limiting down into `MatrixHttpApi` so that it affects _all_ requests?

Reply:
probably just here for now - putting it at the MatrixHttpApi layer is probably going to cause unexpected problems.